### ISSUE TITLE
type-mismatch false positive when early return from match

### ISSUE THREAD

**Issue by jhgg** [rust-analyzer/rust-analyzer#11814]
consider the following code:

```rust
fn example() -> bool {
    match 1 {
        _ => return true,
    };
}
```

This shows the error diagnostic `expected bool, found () (rust-analyzer)[type-mismatch]`
